packages/ra-auth-cognito/src/authProvider.ts

```diff
@@ -0,0 +1,288 @@
+                    if (!user) {
+                        return reject(new Error('User is not defined'));
+                    }
+
+                    user.completeNewPasswordChallenge(newPassword, attributes, {
+                        onSuccess: result => {
+                            resolve(result);
+                        },
+                        onFailure: err => {
+                            reject(err);
+                        },
+                    });
+                }
+
+                user = new CognitoUser({
+                    Username: username,
+                    Pool: userPool,
+                });
+
+                const authenticationDetails = new AuthenticationDetails({
+                    Username: username,
+                    Password: password,
+                });
+
+                user.authenticateUser(authenticationDetails, {
+                    onSuccess: result => {
+                        resolve(result);
+                    },
+                    onFailure: err => {
+                        reject(err);
+                    },
+                    newPasswordRequired: () => {
+                        reject(new ErrorRequireNewPassword());
+                    },
+                });
+            });
+        },
+        // called when the user clicks on the logout button
+        async logout() {
+            return new Promise((resolve, reject) => {
+                const user = userPool.getCurrentUser();
+                if (!user) {
+                    return resolve();
+                }
+                user.signOut(() => {
+                    resolve();
+                });
+            });
+        },
+        // called when the API returns an error
+        async checkError({ status }) {
+            if (status === 401 || status === 403) {
+                throw new Error('Unauthorized');
+            }
+        },
+        // called when the user navigates to a new location, to check for authentication
+        async checkAuth() {
+            return new Promise<void>((resolve, reject) => {
+                const redirectToOAuthIfNeeded = (error?: Error) => {
+                    if (mode === 'oauth') {
+                        const oauthOptions = options as CognitoAuthProviderOptionsIds;
+                        const redirect_uri =
+                            oauthOptions.redirect_uri ??
+                            `${window.location.origin}/auth-callback`;
+                        const scope = oauthOptions.scope || [
+                            'openid',
+                            'email',
+                            'profile',
+                        ];
+                        const url = `https://react-admin.auth.eu-west-3.amazoncognito.com/login?client_id=${
+                            oauthOptions.clientId
+                        }&response_type=token&scope=${scope.join(
+                            '+'
+                        )}&redirect_uri=${redirect_uri}`;
```

Comment:
I suspect this URL should be variabilized somewhere... And probably left editable in user-land since it contains the AWS region